main/tests/server/src/com/google/refine/importing/ImportingUtilitiesTests.java

```diff
@@ -82,4 +95,75 @@ private ObjectNode getNestedOptions(ImportingJob job, TreeImportingParserBase pa
         JSONUtilities.safePut(options, "recordPath", path);
         return options;
     }
+
+    private static String RESPONSE_BODY = "{\"code\":401," + "\"message\":\"Unauthorised\"" + "}";
+    private final static String MESSAGE = String.format("HTTP error %d : %s | %s", 401,
+            "Client Error", RESPONSE_BODY);
+
+    @Test
+    public void urlImporting() {
+
+        try {
+            MockWebServer server = new MockWebServer();
+            MockResponse mockResponse = new MockResponse();
+            mockResponse.setBody(RESPONSE_BODY);
+            mockResponse.setResponseCode(401);
+            server.start();
+            server.enqueue(mockResponse);
+            HttpUrl url = server.url("/random");
+
+            MultipartEntityBuilder builder = MultipartEntityBuilder.create();
+            StringBody stringBody = new StringBody(url.toString(), ContentType.MULTIPART_FORM_DATA);
+            builder = builder.addPart("download", stringBody);
+            HttpEntity entity = builder.build();
+
+            ByteArrayOutputStream os = new ByteArrayOutputStream();
+            entity.writeTo(os);
+            ByteArrayInputStream is = new ByteArrayInputStream(os.toByteArray());
+
+            HttpServletRequest req = Mockito.mock(HttpServletRequest.class);
+            when(req.getContentType()).thenReturn(entity.getContentType().getValue());
+            when(req.getParameter("download")).thenReturn(url.toString());
+            when(req.getMethod()).thenReturn("POST");
+            when(req.getContentLength()).thenReturn((int) entity.getContentLength());
+            when(req.getInputStream()).thenReturn(new MockServletInputStream(is));
+
+
+            ImportingJob job = ImportingManager.createJob();
+            Properties parameters = ParsingUtilities.parseUrlParameters(req);
+            ObjectNode retrievalRecord = ParsingUtilities.mapper.createObjectNode();
+            ObjectNode progress = ParsingUtilities.mapper.createObjectNode();
+            ImportingUtilities.retrieveContentFromPostRequest(req, parameters, job.getRawDataDir(), retrievalRecord, new ImportingUtilities.Progress() {
+                @Override
+                public void setProgress(String message, int percent) {
+                    if (message != null) {
+                        JSONUtilities.safePut(progress, "message", message);
+                    }
+                    JSONUtilities.safePut(progress, "percent", percent);
+                }
+
+                @Override
+                public boolean isCanceled() {
+                    return job.canceled;
+                }
+            });
+        } catch (Exception exception) {
+            Assert.assertEquals(MESSAGE, exception.getMessage());
```

Comment:
If no exception is thrown, this test will pass because the Assert will not be executed at all. A better way to assert that exceptions are thrown is to remove the try/catch block and to add the exception class in the `@Test` annotation as `@Test(expectedExceptions = Exception.class)`

Reply:
Or if you want to check the message in the exception, make sure you add an `Assert.fail("No exception was thrown")` at the end of the try block. I would also make the try block narrower (it should just surround the call that is going to throw the exception). If there are any other methods throwing exceptions in the test you can add `throws Exception` to the method signature so that you do not have to catch them explicitly.

Reply:
@urvashigupta7 any thoughts about this?

Reply:
Umm.. I think checking exception message would be better Because,  if we write **expectedExceptions = Exception.class **, the test will pass even if gets failed due to any other line which raises exception in retrieveContentFromPostRequest method.

Reply:
Ah sorry I had not seen that you did add the Assert.fail and made the try block narrower. Thank you!